Sources/Resource.swift

```diff
@@ -25,134 +25,32 @@
-    }
-
-    var update: Responder = BasicResponder { _ in
-        return Response(status: .methodNotAllowed)
-    }
-
-    public func update(middleware: Middleware..., respond: RespondWithId) {
-        let responder = BasicResponder { request in
-            return try respond(request: request, id: request.id!)
-        }
-        self.update = middleware.chain(to: responder)
-    }
-
-    public func update(middleware: Middleware..., respond: RespondWithIdContent) {
-        let responder = BasicResponder { request in
-            guard let content = request.content else {
-                return Response(status: .badRequest)
-            }
-            return try respond(request: request, id: request.id!, content: content)
-        }
-        self.update = middleware.chain(to: responder)
-    }
-
-    public func update<T: ContentMappable>(middleware: Middleware..., content: T.Type, respond: (request: Request, id: String, content: T) throws -> Response) {
-        let contentMapper = ContentMapperMiddleware(mappingTo: content)
-        let responder = BasicResponder { request in
-            let content = request.storage[T.key] as! T
-            return try respond(request: request, id: request.id!, content: content)
-        }
-        let middlewareChain: [Middleware] = [contentMapper] + middleware
-        self.update = middlewareChain.chain(to: responder)
-    }
-
-    var destroy: Responder = BasicResponder { _ in
-        return Response(status: .methodNotAllowed)
-    }
-
-    public func destroy(middleware: Middleware..., respond: RespondWithId) {
-        let responder = BasicResponder { request in
-            return try respond(request: request, id: request.id!)
-        }
-        self.destroy = middleware.chain(to: responder)
-    }
-}
-
-public struct Resource {
+public struct Resource: HTTP.Router {
     public let middleware: [Middleware]
-    public let index: Responder
-    public let create: Responder
-    public let show: Responder
-    public let update: Responder
-    public let destroy: Responder
+    public let routes: [Route]
+    public let fallback: Responder
+    public let matcher: RouteMatcher
 
-    public init(middleware: Middleware..., mediaTypes: [MediaType], build: (resource: ResourceBuilder) -> Void) {
-        let builder = ResourceBuilder()
+    public init(_ path: String = "", middleware: [Middleware] = [], mediaTypes: MediaType..., build: (resource: ResourceBuilder) -> Void) {
+        let builder = ResourceBuilder(path: path)
         build(resource: builder)
+
         let contentNegotiaton = ContentNegotiationMiddleware(mediaTypes: mediaTypes)
         self.middleware = [contentNegotiaton] + middleware
-        self.index = builder.index
-        self.create = builder.create
-        self.show = builder.show
-        self.update = builder.update
-        self.destroy = builder.destroy
+
+        self.fallback = builder.fallback
+        self.matcher = TrieRouteMatcher(routes: builder.routes)
```

Comment:
I think it would be a nice feature to allow injection of matchers. It would require some modifications to the `RouterMatcher` to provide information about its component splitting and parameter styles, but I can imagine how it would be useful in the future.